### ISSUE TITLE
LDA binary classification is broken

### ISSUE THREAD

**Issue by micmn** [shogun-toolbox/shogun#3886]
This is not due to the last patch but true also for Shogun 6.

Check the notebook on shogun cloud (in fact this build should be prior to the patch) [lda.pdf](https://github.com/shogun-toolbox/shogun/files/1112707/lda.pdf)

I'll look into it asap.

**Comment by thhung** [comment]
I have an problem with the offical shogun on fedora 23 is that after trainning lda always return -1. Is this the same problem? Thanks. 
P/s: sorry for beginner question.

**Comment by micmn** [comment]
Probably, can you share the code you are using?

**Comment by karlnapf** [comment]
This is now fixed as #3901 was merged, right? @micmn

**Comment by thhung** [comment]
I don't have the code with me now but I will post it here tomorrow when I get it. 
The version shogun I'm using is from official shogun package on Fedora 23.

**Comment by thhung** [comment]
Hello, 
Please find below my code: 

Training:

     //Given two mat object of opencv
     cv::Mat trainSamples,  trainLabels;
     //convert label from 0 to -1 
     trainLabels.setTo(-1, trainLabels == 0);
     shogun::init_shogun_with_defaults();

    #define START 0
    #define STOP trainSamples.rows

    SGVector< float64_t > lab(STOP-START);
    SGMatrix< float64_t > feat(trainSamples.cols, STOP - START);
    // It's quite surprise the format of this feat matrix is reversed with format from OpenCV.   
    for (int r = START; r < STOP ; ++r) //trainSamples.rows
        for (int c = 0; c < trainSamples.cols; ++c)
        {
            feat(c,r-START) = (double)(trainSamples.at<float>(r - START,c)); 
        }

    for( int i = START ; i < STOP ; ++i ) // trainLabels.rows
        lab[i - START] = double(trainLabels.at<int>(i - START,0));

    // Create train labels
    CBinaryLabels* labels = new CBinaryLabels(lab);//CMulticlassLabels

    // Create train features
    CDenseFeatures< float64_t >* features = new CDenseFeatures< float64_t >(feat);

     //make sure my labels and features are read correctly
     features->get_feature_matrix().display_matrix();
     labels->get_labels().display_vector();
   
    //simple training
     CLDA* lda = new CLDA(0, SVD_LDA);
     lda->set_features(features);
     lda->set_labels(labels);
    SG_REF(lda);
    lda->train();
    // evaluate training to get training accuracy
    CBinaryLabels* pred = lda->apply_binary(features);
     CAccuracyMeasure evaluate = CAccuracyMeasure();
     evaluate.evaluate(pred, labels);
     EXPECT_GT(evaluate.get_accuracy(), 0.9);

    // This settings I get 92% but there is error in prediction later. 

    CSerializableAsciiFile *file = new CSerializableAsciiFile(argv[2],'w');
    lda->save_serializable(file);
    file->close();

    SG_UNREF(file);
    SG_UNREF(lda);

    shogun::exit_shogun();

    return 0;
    //end the training code.



The testing code:

     init_shogun_with_defaults();
    CSerializableAsciiFile *file = new CSerializableAsciiFile(argv[1],'r');
    CLDA* lda = new CLDA();
    if(!lda->load_serializable(file))
    {
        std::cout << "Error : Can't load model. Quit now!" << std::endl;
        return 1;
    }
      //make sure model load correctly
      lda->print_serializable(); //do you have docs to explain this?
      std::cout << "weights: " << std::endl;
      lda->get_w().display_vector(); //weights are correct.
      file->close();
     //Given mat object as features
      cv::Mat mat_features;

      SGMatrix< float64_t > feat(mat_features.cols, mat_features.rows);

       for (int r = 0; r < mat_features.rows ; ++r) 
          for (int c = 0; c < mat_features.cols; ++c)
         {
             feat(c,r) = (double)(mat_features.at<float>(r,c));
         }

      CDenseFeatures< float64_t >* features = new CDenseFeatures< float64_t >(feat);

       CBinaryLabels* pred = lda->apply_binary(features);
    
      //here I get all -1
       pred->get_labels().display_vector();

     exit_shogun();
     return 0;
    //end testing code.


That's my code for training and testing. Do you see something I did wrong? Thanks for your help.

**Comment by vigsterkr** [comment]
@thhung there's actually an opencv <-> shogun helper function in shogun.. only thing is that it's not compiled with the official fedora package.

**Comment by vigsterkr** [comment]
@thhung but on the other hand just recently has been some serious bug fixed in shogun's LDA so i recommend you to try to compile shogun from source... in that case you should use `-DOpenCV=On` flag for `cmake` and check the docs here: https://github.com/shogun-toolbox/shogun/tree/develop/doc/OpenCV_docs

**Comment by micmn** [comment]
@thhung can you check also the bias value in both training/testing (lda->get_bias())? that could be the problem and it should be fixed in latest version

**Comment by thhung** [comment]
Thanks for your help! I built from source with the master branch. But it didn't work. It always returns -1.
Here is my bias: 
   training: -0.00175954
   testing : -0.00175954

It's kind of strange because when I train the same dataset in scikit-learn or matlab ( all default parameters), it's ok. But here I can't reproduce it here.

**Comment by micmn** [comment]
You should try develop branch https://github.com/shogun-toolbox/shogun/tree/develop

Also, if you can share the data I can try to reproduce it.

**Comment by thhung** [comment]
Thanks, I will try it. 
[dummy_test.csv.txt](https://github.com/shogun-toolbox/shogun/files/1128379/dummy_test.csv.txt)
[dummy_train.csv.txt](https://github.com/shogun-toolbox/shogun/files/1128380/dummy_train.csv.txt)
Please find above my data. I will try the develop.

**Comment by micmn** [comment]
Is this the correct data? dummy_test seems 0s, NaNs and garbage.

**Comment by thhung** [comment]
Yeah I see.I didn't expect that. I couldn't connect to my machine now. But I will post new data here as soon as possible tomorrow. It's strange.

**Comment by thhung** [comment]
Thanks a bunch, guys. I found the problem with the encode of my Mat object (I men  the type). Finally I can reproduce the result of scikit-learn with shogun-master. By the way, what is the different of LDA between master and develop branch. Because I only tried with master branch.

**Comment by karlnapf** [comment]
The one in master branch has a bug (might give wrong results in some cases)
The one in develop branch was recently updated, fixing all errors, and running faster than sklearn now ...
